### PR TITLE
Adds functions for sampling nodes from a graph

### PR DESCRIPTION
**Pull request by jfinkels**
This pull request includes code from issue #463. I'm not sure how to properly test randomized sampling functions, though.

This module may also be a good place for functions that sample from trees, see issue #1125.

### CODE REVIEW

File: networkx/algorithms/sampling.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+import itertools
+import random
+
+import networkx as nx
+
+__all__ = ['degree_weighted_independent_node_sample', 'random_walk',
+           'uniform_independent_node_sample']
+
+
+# def estimate_mean(sample, values, weights=None):
```
Comment (ysitu) [inline networkx/algorithms/sampling.py]: Why commented out?

File: networkx/algorithms/sampling.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+            *INFOCOM, 2010 Proceedings IEEE*. IEEE, 2010.
+            http://dx.doi.org/10.1109/INFCOM.2010.5462078
+
+    """
+    if weights is None:
+        for c in itertools.count():
+            if c == size:
+                return
+            yield random.choice(population)
+    cum_weights = nx.utils.accumulate(weights(v) for v in population)
```
Comment (ysitu) [inline networkx/algorithms/sampling.py]: Isn't `sum` good enough?

File: networkx/algorithms/sampling.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+
+    Examples
+    --------
+    Sampling twenty nodes uniformly at random from the wheel graph on ten
+    vertices::
+
+        >>> import networkx as nx
+        >>> G = nx.generators.wheel_graph(10)
+        >>> list(uniform_independent_node_sample(G, size=20))
+        [8, 6, 4, 7, 9, 5, 5, 2, 4, 5, 7, 9, 3, 6, 1, 6, 9, 8, 6, 4]
```
Comment (ysitu) [inline networkx/algorithms/sampling.py]: You need to make doctest skip this.

File: networkx/algorithms/sampling.py [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+    """
+    return sample_with_replacements(G, size)
```
Comment (ysitu) [inline networkx/algorithms/sampling.py]: The first argument to `sample_with_replacements` should be a `list` as documented.

File: networkx/algorithms/sampling.py (hunk 1 above)
Comment (jfinkels) [inline networkx/algorithms/sampling.py]: I question whether this should function should be provided by NetworkX. This function has nothing in particular to do with graphs, so it should not be provided as part of the public API of NetworkX. At best, these should be provided as examples only.

File: networkx/algorithms/sampling.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+       to be sampled.
+
+    Returns
+    -------
+    iterator
+       An iterator over elements of ``population``.
+
+    References
+    ----------
+    .. [1]: Gjoka, Minas, Maciej Kurant, Carter T. Butts,
```
Comment (ysitu) [inline networkx/algorithms/sampling.py]: You mean "Minas Gjoka"?

Comment (jfinkels) [inline networkx/algorithms/sampling.py]: MLA style, for example, inverts the name order of the first author. Search for the paper on Google Scholar, then click the "Cite" link to see some examples.

File: networkx/algorithms/sampling.py (hunk 2 above)
Comment (jfinkels) [inline networkx/algorithms/sampling.py]: No, the cumulative sums are used in the bisection below.

File: networkx/algorithms/sampling.py (hunk 3 above)
Comment (jfinkels) [inline networkx/algorithms/sampling.py]: How do I do that?

File: networkx/algorithms/sampling.py (hunk 2 above)
Comment (ysitu) [inline networkx/algorithms/sampling.py]: My bad. I misread this.

File: networkx/algorithms/sampling.py (hunk 1 above)
Comment (hagberg) [inline networkx/algorithms/sampling.py]: I think it is questionable whether any of these functions belong in the
public networkx API.

Comment (jfinkels) [inline networkx/algorithms/sampling.py]: Also a fair question, it should probably discussed (directly on the pull request).

**Comment by hagberg** [comment]
There are some random sample helpers in networkx.utils too.

**Comment by jfinkels** [comment]
I'll take a look at combining those with this code.

**Comment by hagberg** [comment]
Maybe they should be in "utils" or as examples?
There is also this old thread on the mailing list about choosing random edges 
https://groups.google.com/forum/#!topic/networkx-discuss/qH3WynBjp3g
Maybe something useful there?

File: networkx/algorithms/sampling.py (hunk 4 above)
Comment (jfinkels) [inline networkx/algorithms/sampling.py]: Right, I will fix that.

**Comment by jfinkels** [comment]
@hagberg wrote:

> There are some random sample helpers in networkx.utils too.

I looked at the `random_*` functions in `networkx.utils`, but they were not particularly applicable to the task of sampling nodes or edges from a graph.

> Maybe they should be in "utils" or as examples?

As I proposed in #1533, only internal functions should live in a `utils` package; if these functions are not used elsewhere in the code as helper functions, they should not be put in a `utils` package.

There is, however, an argument to made for these to be examples only. I'm not making that argument, though, so perhaps someone else can. I think these should be part of the public API; the issues I referenced originally indicate some interest in sampling.

> There is also this old thread on the mailing list about choosing random edges
> https://groups.google.com/forum/#!topic/networkx-discuss/qH3WynBjp3g
> Maybe something useful there? 

I added a modified version of the edge sampling code from your post there and updated the pull request. One problem is that one of the functions you provided comes from ActiveState, which has a silly and burdensome copyright license requirement, as I describe at the head of the module.

**Comment by jfinkels** [comment]
Okay, I've updated this pull request so there are three major modules in the `sampling` package, one for sampling nodes, one for sampling edges, and one for generating random walks. A fourth module contains the generic sampling methods. It also has unit tests now.

I removed the code for sampling with replacement from the Google groups message reference above because it didn't apply to sample sizes greater than the size of the population. I removed the code for estimating the relative size of a set given a sample of a population, but someone else might make a pull request for them in a separate `algorithms.statistics` module, say. They now live in a gist [here][1].

[1]: https://gist.github.com/jfinkels/f304a285098e9a4606bd

**Comment by dschult** [comment]
I'm commenting here based on a chat comment from yesterday's community meeting.

This is an 8 year old PR that adds sampling functions for edges and nodes. Basically, the functions let you select a set of N random nodes (or random edges). Most are chosen uniformly random, but there's also functionality to prescribe probabilities to each.

I think this can be done now with something like 
```python
rng = np.random.default_rng()
sample_nodes = rng.choice(G.nodes, size=15, p=[n/sum(G.nodes) for n in G.nodes])
```

So perhaps this should be closed -- or maybe find a place in the docs to point this out? Or suggest a nx-guide topic on this?  Do we have a place to suggest nx-guide topics?  Maybe in the mentored projects?

**Comment by dschult** [comment]
I'm going to close this issue. I think sampling interfaces  generally and the NX random walks functions make this easier.
Comment here if you are still interested in this.

**Pull request closed without merging**